### PR TITLE
feat: add histogram support to service metrics

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Add histograms to service metrics.
Add histogram significant digit setting. (default 5)
Update tests.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Closes https://github.com/elastic/apm-server/issues/9656

### CODE REVIEW

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 1]
```diff
@@ -268,13 +291,23 @@ func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetric
...
-		transactionDuration: metrics.transactionDuration,
-		transactionCount:    metrics.transactionCount,
-		failureCount:        metrics.failureCount,
-		successCount:        metrics.successCount,
+	if entry.serviceMetrics.histogram == nil {
+		entry.serviceMetrics = serviceMetrics{
+			transactionDuration: metrics.transactionDuration,
+			transactionCount:    metrics.transactionCount,
+			failureCount:        metrics.failureCount,
+			successCount:        metrics.successCount,
```
Comment (simitt) [inline x-pack/apm-server/aggregation/servicemetrics/aggregator.go:299]: Why would you set `duration`, `count`, `failureCount` and `successCount` dependent of `histogram==nil`? Isn't that independent of the histogram initialization?

**Review by simitt — COMMENTED** [review]
Generally looks good; afaics the same defaults are used as in the transaction metrics aggregation. Left one comment.

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go (hunk 1 above)
Comment (kruskall) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go]: Good catch! Updated! +1

File: x-pack/apm-server/aggregation/servicemetrics/aggregator.go [hunk 2]
```diff
@@ -381,6 +442,17 @@ func makeMetricset(key aggregationKey, metrics serviceMetrics) model.APMEvent {
 				Count: metricCount,
 				Sum:   float64(time.Duration(math.Round(metrics.transactionDuration)).Microseconds()),
 			},
+			DurationHistogram: model.Histogram{
+				Counts: counts,
+				Values: values,
+			},
```
Comment (lahsivjar) [inline x-pack/apm-server/aggregation/servicemetrics/aggregator.go:448]: Now that we have histograms, do we need to keep the `SummaryMetric`?

Comment (kruskall) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go:448]: Good question! I think so but my assumption is based on the output document @axw posted in the kibana issue: https://github.com/elastic/kibana/issues/146734#issuecomment-1342467414

Comment (axw) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go:448]: Yes we still need them, in fact we'll be adding them to transaction metrics too: https://github.com/elastic/apm-server/issues/9655

The reason for having both is that calculating sum, count, and average over histograms is considerably more expensive due to the way they're stored. Elasticsearch ends up having to run through all the buckets of the histogram, vs. looking up one or two sub-fields of an `aggregate_metric_double`.

In the future we should be able to store a histogram along with sum, count, and optionally min & max in a single field: https://github.com/elastic/elasticsearch/issues/81031

Comment (lahsivjar) [reply x-pack/apm-server/aggregation/servicemetrics/aggregator.go:448]: Ah, that makes sense. Thanks for the clarification, Andrew.

**Review by lahsivjar — APPROVED** [review]
LGTM!

**Review by simitt — APPROVED** [review]

**Comment by marclop** [comment]
Tested with 8.7.0 BC2 on ESS:

```go
package main

import (
	"fmt"
	"time"

	"go.elastic.co/apm/v2"
)

func main() {
	tracer := apm.DefaultTracer()
	tracer.SetSpanCompressionEnabled(false)             // Disable compression.
	tracer.SetExitSpanMinDuration(2 * time.Millisecond) // set duration to 2ms.
	for i := 0; i < 3_000; i++ {
		for _, duration := range []int{1, 100} {
			tx := tracer.StartTransaction(fmt.Sprintf("tx-%d", i), fmt.Sprintf("type-%d", i))
			// This span will get dropped since it's 1ms and the minimum span
			// duration to report is 2ms.
			span := tx.StartSpanOptions(fmt.Sprintf("span-%d", i), fmt.Sprintf("type-%d", i), apm.SpanOptions{
				ExitSpan: true,
			})
			span.Duration = time.Duration(duration) * time.Millisecond
			span.End()
			tx.Duration = time.Duration(duration) * time.Millisecond
			tx.End()
		}
		if i%500 == 0 {
			tracer.Flush(nil)
		}
	}
	tracer.Flush(nil)
	fmt.Printf("%+v\n", tracer.Stats())
}
```

```
GET /metrics-apm*/_search?filter_path=hits.hits
{
  "size": 2, 
  "_source": [false], 
  "fields": [
    "metricset.name",
    "transaction.duration.histogram"
  ], 
  "query": {
    "match": {
      "metricset.name": "service_transaction"
    }
  }
}
```

```
{
  "hits": {
    "hits": [
      {
        "_index": ".ds-metrics-apm.service_transaction.1m-default-2023.02.20-000001",
        "_id": "w6vpbIYBZ-vqtropJh-E",
        "_score": 1,
        "_source": {},
        "fields": {
          "metricset.name": [
            "service_transaction"
          ],
          "transaction.duration.histogram": [
            {
              "values": [
                1003,
                100351
              ],
              "counts": [
                1,
                1
              ]
            }
          ]
        }
      },
      {
        "_index": ".ds-metrics-apm.service_transaction.1m-default-2023.02.20-000001",
        "_id": "xKvpbIYBZ-vqtropJh-E",
        "_score": 1,
        "_source": {},
        "fields": {
          "metricset.name": [
            "service_transaction"
          ],
          "transaction.duration.histogram": [
            {
              "values": [
                1003,
                100351
              ],
              "counts": [
                1,
                1
              ]
            }
          ]
        }
      }
    ]
  }
}
```

**Pull request merged**